driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -168,117 +166,79 @@ public void run() {
-                final List<Connection> connections =
-                    manager.connectionFactory().newConnections(HostConnectionPool.this, needed);
-                for (Connection connection : connections) {
-                  ListenableFuture<Void> connectionFuture = connection.initAsync();
-                  connectionFutures.add(handleErrors(connectionFuture, initExecutor));
-                }
-                addCallback(connections, connectionFutures, initFuture, toClose);
-              } else {
-                for (final Connection c : toClose) {
-                  c.closeAsync();
-                }
-                phase.compareAndSet(Phase.INITIALIZING, Phase.READY);
-                initFuture.set(null);
-              }
-            }
-          }
-
-          @Override
-          public void onFailure(Throwable t) {
-            phase.compareAndSet(Phase.INITIALIZING, Phase.INIT_FAILED);
-            forceClose(connections);
-            forceClose(toClose);
-            for (List<Connection> shardConnections : HostConnectionPool.this.connections) {
-              forceClose(shardConnections);
-            }
-            open.set(0);
-            initFuture.setException(t);
-          }
-        },
-        initExecutor);
-  }
+		final Connection connection,
+		final ListenableFuture<Void> connectionFuture,
+		final SettableFuture<Void> initFuture,
+		final int count) {
+
+		final Executor initExecutor = manager.cluster.manager.configuration.getPoolingOptions()
+				.getInitializationExecutor();
+	
+		GuavaCompatibility.INSTANCE.addCallback(connectionFuture, new FutureCallback<Void>() {
+			@Override
+			public void onSuccess(Void l) {
+				int shardId = connection.shardId();
+				if (isClosed()) {
+					initFuture.setException(new ConnectionException(host.getSocketAddress(),
+							"Pool was closed during initialization"));
+					// we're not sure if closeAsync() saw the connections, so ensure they get closed
+					connection.closeAsync().force();
+					scheduledForCreation[shardId].decrementAndGet();
+				} else {
+					if (!connection.isClosed()) {
+						synchronized(connections[shardId]) {
+							if (connections[shardId].size() < connectionsPerShard) {
+								scheduledForCreation[shardId].decrementAndGet();
+								open.incrementAndGet();
+								HostConnectionPool.this.connections[shardId].add(connection);
+								if (phase.compareAndSet(Phase.INITIALIZING, Phase.READY)) {
+									initFuture.set(null);
+								}
+								return;
+							} else {
+								connection.closeAsync();
+							}
+						}
+					}
+					// Re-attempt to establish this connection
+					Connection newConnection = manager.connectionFactory().newConnections(HostConnectionPool.this,1).get(0);
+					addCallback(newConnection,
+							handleErrors(newConnection.initAsync(), initExecutor),
+							initFuture,
+							count);
+				}
+			}
+
+			@Override
+			public void onFailure(Throwable t) {
+				int f = failed.incrementAndGet();
```

Comment:
I think we should try to connect again if a connection failed. At least some number of times.

Reply:
The code does that. Notice that the onFailure callback is only invoked when a non-recoverable exception is thrown in the Connection class. In all other cases, the onSuccess is invoked with a closed connection, which we re-open.

Reply:
I know that closed connection is being re-opened but non-recoverable exception does not mean we can't open a new connection instead of this one that failed.

Reply:
No, it means exactly that. The exceptions that lead to onFailure are: ClusterNameMismatchException, UnsupportedProtocolVersionException and AuthenticationException. You cannot recover from that. Check the handleErrors() method.

Reply:
Why are you counting the failures then, instead of terminating the init process after the first non-recoverable exception?

Reply:
Because that was the former behavior, which I kept. But you're right, we can fail altogether at the first onFailure execution.

Reply:
The former behaviour was that the whole init process of the pool was failing when a single connection got non-recoverable exception. Not waiting for other connections to finish.